File: TP2/Exo 2.6/python/phrases.py
def main():
    # Demander à l'utilisateur de saisir le nom du fichier
    filename = input("Entrez le nom du fichier : ")
    
    # Demander à l'utilisateur de saisir la phrase à rechercher
    phrase = input("Entrez la phrase à rechercher : ")
    
    try:
        # Ouvrir le fichier et lire son contenu
        with open(filename, 'r', encoding='utf-8') as file:
            lines = file.readlines()  # Lire toutes les lignes du fichier dans une liste

        # Initialiser le numéro de ligne
        line_number = 0
        
        # Rechercher la phrase dans chaque ligne
        for line in lines:
            line_number += 1  # Incrémenter le compteur de ligne
            count = line.count(phrase)  # Compter le nombre d'occurrences de la phrase dans la ligne

            # Afficher le résultat si la phrase est trouvée dans la ligne
            if count > 0:
                print(f"Ligne {line_number} : {count} occurrence(s) de '{phrase}'")

    # Gestion de l'erreur si le fichier n'est pas trouvé
    except FileNotFoundError:
        print("Erreur : impossible d'ouvrir le fichier.")

# Point d'entrée du programme
if __name__ == "__main__":
    main()

File: TP2/Exo 2.6/C/phrases.c
#include <stdio.h>
#include <string.h>

// Définition de la longueur maximale d'une ligne
#define MAX_LINE_LENGTH 256 

int main() {
    char filename[100]; // Variable pour stocker le nom du fichier
    char phrase[100];   // Variable pour stocker la phrase à rechercher
    FILE *file;         // Pointeur vers le fichier
    char line[MAX_LINE_LENGTH]; // Variable pour stocker chaque ligne lue
    int line_number = 0; // Compteur de lignes
    int count;           // Compteur d'occurrences de la phrase dans une ligne

    // Demander à l'utilisateur de saisir le nom du fichier
    printf("Entrez le nom du fichier : ");
    scanf("%s", filename);

    // Demander à l'utilisateur de saisir la phrase à rechercher
    printf("Entrez la phrase à rechercher : ");
    scanf(" %[^\n]s", phrase); // Lit une phrase complète jusqu'à la nouvelle ligne

    // Ouvrir le fichier en lecture
    file = fopen(filename, "r");
    if (file == NULL) {
        // Si le fichier ne peut pas être ouvert, afficher une erreur
        printf("Erreur : impossible d'ouvrir le fichier.\n");
        return 1; // Retourne 1 pour indiquer une erreur
    }

    // Lire le fichier ligne par ligne
    while (fgets(line, sizeof(line), file) != NULL) {
        line_number++; // Incrémenter le compteur de lignes
        count = 0;     // Réinitialiser le compteur d'occurrences pour chaque ligne

        // Compter le nombre de fois que la phrase apparaît dans la ligne
        char *ptr = line; // Pointeur pour parcourir la ligne
        while ((ptr = strstr(ptr, phrase)) != NULL) {
            count++;                 // Incrémenter le compteur d'occurrences
            ptr += strlen(phrase);   // Avancer le pointeur pour rechercher la prochaine occurrence
        }

        // Afficher le résultat si la phrase est trouvée dans la ligne
        if (count > 0) {
            printf("Ligne %d : %d occurrence(s) de '%s'\n", line_number, count, phrase);
        }
    }

    // Fermer le fichier après la lecture
    fclose(file);
    return 0; // Retourne 0 pour indiquer que tout s'est bien passé
}
